### DISCUSSION TITLE
List files in zip ?

### DISCUSSION THREAD

**Discussion by CouinCouin** [wqweto/ZipArchive#27] [General]
Hi wqweto :) 

In my project, when I launch the "Backup" function, I zip files and classing them in subfolders based on CSIDLs ( related with Elroy's message [here](https://www.vbforums.com/showthread.php?898444-Error-5-Bad-filename-or-number-when-trying-to-Dir-a-path&p=5587609&viewfull=1#post5587609) ), so for example, a file on user Music's folder will be ziped in CSIDL/13/MyFile.mp3 . It is usefull to restore it on the current user Music's folder of another machine, especially if the user name is different and is we pass from XP <=> Win 10 (avoiding the problem "Users" folder instead of "Documents and settings"). All files are designed from a .ini where all used files (by software) are listed. Of course, I include this ini file in the archive.

To "Restore" , I unzip the ini file, and after, I read the ini file to extract files from archive.

Even if the unzip script looks ok in my project, I'm curious and I would know if there way to list (recursivly)  files that are stored in a zip file , with your library ?

Thanks :)
Couin

**Comment by wqweto** [comment]
Hi, 

> To "Restore" , I unzip the ini file, and after, I read the ini file to extract files from archive.

You can extract this INI file in-memory without writing any file to disk like this
```
Dim baOutput() As Byte
bResult = .Extract(baOutput, "list_of_files.ini")
Debug.Print StrConv(baOutput, vbUnicode)
```
. . . where `list_of_files.ini` is the name of the INI file.

For enumerating all the entries (files) inside the ZIP you can use `FileCount` and `FileInfo` properties like this
```
    Dim lIdx            As Long
    
    With New cZipArchive
        .OpenArchive "D:\Temp\aaa8.zip"
        For lIdx = 0 To .FileCount - 1
            Debug.Print "FileName=" & .FileInfo(lIdx, zipIdxFileName) & ", Size=" & .FileInfo(lIdx, zipIdxSize)
        Next
    End With
```
Note that there was a bug in `FileInfo` that I just fixed in 623f34f9d41223af4ae80cecb5bee80cb3b78a07 so you'll need to get latest version of the repo.

**Comment by CouinCouin** [comment]
Hi , 

Thanks :) 
Hmm, for the ini file, If I extract to Debug?print, The app would not able to read the ini file after ?  Or perhaps there is way to read the Debug console (when running exe, not in VB6 IDE I mean) . 
Also, the ini extracting the ini file on the disk is not a problem, as well as I kremove it once restoration finished. 

About listing files, thanks, I get some result :) 
But I just noticed that for files in CSIDL , there is one line per file, but for "W", there is one line per subfolder and then a line for the file. 

> FileName=CSIDL\16\Nexus 8 - Midnight express (Intercity mix).aiff, Size=61424610
> FileName=CSIDL\39\West_Musette.mp3, Size=71889
> FileName=CSIDL\13\Yeee_Haaa.mp3, Size=30354
> FileName=W\jp\, Size=0
> FileName=W\jp\samples\, Size=0
> FileName=W\jp\samples\poc\, Size=0
> FileName=W\jp\samples\poc\Radio_FG_(FR).mp3, Size=383060 

Not really a problem (just have to keep lines where FileName finishes by other than \ ) , but just to know if normal or not ?  

Thanks :)

**Comment by wqweto** [comment]
`W\jp` and `W\jp\samples` are some directories inside your ZIP archive. Directory entries inside the archive are marked with `vbDirectory` attribute set. That's why I said the code enumerates archive entries, not files -- this includes files and directories.

About the INI file: my idea was to read the contents of the INI file to a string with something like this
```
Dim MyIni As String: MyIni = StrConv(baOutput, vbUnicode)
```
. . . not to `Debug.Print` it in your production code exactly.

If this doesn't ring a bell with you then probably you have no use of this INI content reading code so you can ignore it.

**Comment by CouinCouin** [comment]
Oh ok for the ini, I understand (well, I think ) :) 

About deirectories inside zip archive, CSIDL and numbers (subofledr)s are directories too, I was just a little suprised to not have something like this : 
FileName=CSIDL\16\, Size=0
FileName=CSIDL\16\Nexus 8 - Midnight express (Intercity mix).aiff, Size=61424610
FileName=CSIDL\39\, Size=0
FileName=CSIDL\39\West_Musette.mp3, Size=71889
FileName=CSIDL\13\, Size=0
FileName=CSIDL\13\Yeee_Haaa.mp3, Size=30354
FileName=W\jp, Size=0
FileName=W\jp\samples, Size=0
FileName=W\jp\samples\poc, Size=0
FileName=W\jp\samples\poc\Radio_FG_(FR).mp3, Size=383060

But no problem, I adpated my code so lines FileNames that finished by " \ " are not threated :)

**Reply by wqweto** [reply]
Btw, you can test the archive entry's attributes for `vbDirectory` like this
```
For lIdx = 0 To .FileCount - 1
    If (.FileInfo(lIdx, zipIdxAttributes) And vbDirectory) <> 0 Then
        '--- it's a directory
    Else
        '--- it's a file
    End If
Next
```

**Reply by CouinCouin** [reply]
Thanks :)
